Do not relay on Linux cooked-mode capture (SLL)

yashi opened the pull request:
Linux cooked-mode capture (SLL) is the pseudo-protocol used by libpcap on Linux when capturing on the "any" interface.  If you don't use "any", you won't get it.  That means that if you capture on a one CAN device, the dissectors don't work.

It seems SLL is only used for heuristic determination of CSP packets and for the key, to which sll_pkt is always "1:".

Remove sll.pkttype and sll.ltype so that the dissector works on captures without cooked-mode capture.

yashi committed:
Do not relay on Linux cooked-mode capture (SLL)

Linux cooked-mode capture (SLL) is the pseudo-protocol used by libpcap
on Linux when capturing on the "any" interface.  If you don't use
"any", you won't get it.  That means that if you capture on a one CAN
device, the dissectors don't work.

It seems SLL is only used for heuristic determination of CSP packets
and for the key, to which sll_pkt is always "1:".

Remove sll.pkttype and sll.ltype so that the dissector works on
captures without cooked-mode capture.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

KantaTamura commented:
PRありがとうございます！

実装時に放置していた問題でした...下のような理由からより良い解決法を探しています．
何かアイデアはありますか？

## SLLを使用している理由

当初の実装ではSLLを使用していなかったのですが、CANパケットを眺めていると"送信したデータ"と"CANバスを通って戻ってくるデータ"の2種類の同じデータをキャプチャしていることがわかりました．

e.g. `sample/cspv2_can_ping.pcapng` No.1, No.14は全く同じSrc/Dst, Souce countだが、No.1-No.13は送信データ(Send by us)で、No.14-  は受信データ(Broadcast)

そのため、応急的に下記のような区別できるデータを用いて `key` を分けていました．
- 送信データ `sll.pkttype==4 (Send by us)`
- 受信データ `sll.pkttype==1 (Broadcast)`

## 上記の実装を行っている理由

実装上でこのような問題が発生する原因は、CSPの `Fragment Counter` に割り当てられたbit数 (3bit=8種類) が少ないことです．
大きなデータをCSPパケットに載せた場合は `Fragment Counter` が巡回し、パケット単位では何バイト目のデータであるかが分からなくなってしまいます．
そこで、下のコードのように概ね順番通りにパケットが来ることを期待して実装しています．
```
if pinfo.visited == false then
        ...

        local id = csp_frag_cnt
        while not(DataTable[key].fragments[id] == nil) do
            # すでに使用している番地なら8番地ずらす
            id = id + 8
        end
        # データ格納
        DataTable[key].fragments[id] = { data=buffer:bytes(csp_frame_start + extended_header_len, can_len.value - extended_header_len) }
end
```

このように、`Fragment Counter` が巡回した際の処理を行っているため、全く同じ key のデータが来た際に正しく処理できなくなります．

yashi commented:
`sll.pkttype==4 (Send by us)` を capture する方法を教えてください。

CSP ping と `cangen` 両方を試したのですが、 send 側の `can1` でも `ssl.pkttype` が `4` のものは取れませんでした。


- wireshark: 4.2.0
- libpcap: 1.10.4
- Linux: 6.5.13

KantaTamura commented:
sampleディレクトリのデータは下記の環境で取得しています

### 送信側
- wsl ubuntu (linux 5.15.90.4)
- wireshark 3.6.2

### 受信側
- scsat1-rpi

KantaTamura committed:
fix can.id endian

libpcap 1.10.2より、CAN captureにSLLを使用しなくなりました．
それと同時にcan.idに対して行われていたbyte-swapsが行われなくなったため、
ビッグエンディアンで読み取るように変更しています．

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura commented:
WSLのlibpcapのバージョンが古いことが原因でした．
libpcapの変更で `can.id` のエンディアンが変更されていたため対応しました．

ありがとうございます！

ref. 
- https://github.com/the-tcpdump-group/libpcap/issues/1052
